DllImportGenerator/DllImportGenerator/StubCodeContext.cs

```diff
@@ -46,6 +46,10 @@ public enum Stage
 
         public Stage CurrentStage { get; protected set; }
 
+        public abstract bool PinningSupported { get; }
```

Comment:
I think I just need someone to elaborate on the inheritance design aspect of this.

Reply:
The idea I had was that the `StubCodeContext` would be the base type for all of the generators (struct, reverse P/Invoke, P/Invoke, etc) and the subclasses would implement the actual generation. So the P/Invoke generator (the only one we have today) would set both of these to true, but a struct generator would set both to false.

Reply:
That model is surprising when the `StubCodeContext` has a function like `GetIdentifiers()` or `GenerateReturnNativeIdentifier()` which doesn't seem to have a relation of a struct marshaller. I am okay with inheritance but not "just because". Can't the context be passed in? I don't see the obvious win for this architecture and it is starting to crack given the above methods on it - including the constants which also don't seem related to struct marshalling. I could be missing the bigger picture though.

Reply:
`GetIdentifiers` can still make sense in the struct world, but I agree that `GenerateReturnNativeIdentifier` doesn't make sense. I think that `GenerateReturnNativeIdentifier` shouldn't be called from marshallers directly. It should be called from the generator or the behavior should be enabled by default.

Reply:
Then the marshallers would only call `GetIdentifiers` to know which identifier names to use to refer to the managed and native values. A struct code generator could override the implementation of `GetIdentifiers` to return identifiers that point to fields on the managed and native types.

Reply:
I've tried out a refactoring to make the StubCodeContext not have any knowledge of "return identifiers" in the base class. All of that logic now lives in the P/Invoke subclass.